Fix ABP hangup and add new getters

flhofer opened the pull request:
Changed library version to match firmware v1.2.3, updated firmware with fixes to 1.2.3 from pull request [35](https://github.com/arduino/mkrwan1300-fw/pull/35)

I've fixed a problem where the getJoinStatus function waited forever for a response as it parsed the wrong `+OK` response still in the buffer.  Also a minor issue with the channel mask's char buffer.

Added getters for Power, RSSI, and SNR.

flhofer committed:
Update firmware version

flhofer committed:
Library constructor preset defaults

flhofer committed:
Fix sendMask, char buffer must be 5 for hex int, 4+`\n`

flhofer committed:
Add getPower to request last power status

flhofer committed:
Add getRSSI and getSNR

flhofer committed:
Fix Join status request, missed to parse +OK

flhofer committed:
Fix getJoinStatus return type and standardized wait call

flhofer committed:
Changed join status to true false as for FW

flhofer committed:
Updated firmware with fixes from pull request #35

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino-libraries/MKRWAN?pullRequest=93) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino-libraries/MKRWAN?pullRequest=93) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino-libraries/MKRWAN?pullRequest=93) it.</sub>

flhofer committed:
Update MKRWAN (firmware) version number 1.2.4 (w/ fw.h)

giulcioffi commented on src/MKRWAN.h:
```diff
@@ -917,6 +947,10 @@ class LoRaModem : public Stream
     if (waitResponse(timeout, "+EVENT=1,1") != 1) {
       return false;
     }
+    // Second answer, always returns +OK
+    if (waitResponse() != 1) {
+      return false;
+    }
```
Hi @flhofer, I tested you PRs. In ABP everything works fine, but not in OTAA. 
When `modem.joinOTAA(appEui, appKey)` is run in the sketch, the connection actually takes place and I am able to see it in the TTN console. However, these 3 additional lines of code return a failure for OTAA. The result is that even if the OTAA connection happened, the sketch shows a failure.
Removing these 3 lines, I am able to successfully connect both in ABP and OTAA.

flhofer replied:
Hi @giulcioffi 
Which firmware version are you using? Recent versions (>1.2.0) respond with additional `+OK`s. I had the opposite problem where it was waiting forever for a `=` when checking status due to an additional `+OK` without =

flhofer replied:
We might need to decide here. `+OK` everywhere as it is implemented now does not allow discerning if the `+OK` answer is the response to a command or to a query `+OK=`.
My proposal is to change responses to command:value, like I wrote in the other pull request for the firmware.

giulcioffi replied:
I used the firmware from your PR https://github.com/arduino/mkrwan1300-fw/pull/35 :) This issue is encountered only in OTAA. If you could test OTAA and eventually modify the firmware to provide that additional +OK, it would be great!

flhofer replied:
Hi @giulcioffi,
I finally figured where the `+OK` came from and removed it altogether (a library thing issuing two commands in a row). While there, I abstracted getters and setters to common calls, simplifying the library and easing the change to `+<CMD>=<VALUE>` responses. 

This new push now contains the following changes
- added timeout to `skipUntil`
- two common getters for almost all calls
- one setter for almost all calls
- an optimization of waitResponse that
    1. peeks the buffer and checks match only on delimiter hit ('\r','=','+')
    2. allows for parsing of multiple responses in the same string without compromising buffer, e.g. `+EVENT=0,0+OK\r`
    3. parses binary receive `+RECVB=`

- getter response parse match, e.g. `+DR?` responds with `+DR=5` and autodetect based on the version number to accept `+OK=5` too

I tested it for a few hours here, but I will continue tomorrow. Anyone who can join is warmly welcome

Florian

giulcioffi replied:
Hi @flhofer, thanks for your effort!! I compiled the firmware with your latest commit https://github.com/arduino/mkrwan1300-fw/pull/35/commits/f44055655d0cabc5764dc6121d26f4c221d9040c, but I am then unable to start the module. Is this still a work in progress or am I missing something?

flhofer replied:
@giulcioffi yes, still somehow work in progress. For my project, I need almost everything in the library to work seamlessly but I got stuck with the `+RFPOWER` command not working. Anyway, at least now with the latest changes I get a proper response `+ERR_PARAM`, telling me where to look.

This to say, I shared the changes but did not test everything thoroughly. Anyway, I agree let's keep it here and fix everything until there is a certain amount of confidence that it works as expected.

flhofer replied:
@giulcioffi fixed the `start module` issue. It was an inverted return value check in the `begin`, the only setter in the whole code like this 😆 

In the meantime, I found also some other small issues in the firmware like the TXpower set that didn't work. Also, some of the library functions aren't even implemented in the firmware. I added the constants in the library, but obviously, the modem won't do anything.

#define AT_UART		  "+UART"
#define AT_FACNEW	  "+FACNEW"
#define AT_SLEEP	  "+SLEEP"
#define AT_MSIZE	  "+MSIZE"

More tests and updates to come

giulcioffi replied:
Hi @flhofer, thank you again for your contribution! I just tested the new firmware and library in OTAA and ABP and the communication works perfectly :) please keep us updated on your tests so that we can then have a final review.

flhofer replied:
I think this part is done :)

flhofer committed:
Unified setters and getters, with updated +<CMD>=<VALUE> response

Optimized parser of waitResponse that peeks for terminator allowing
chained commands to be parsed. Auto-detect compatibility mode <1.2.4
Includes firmware 1.2.4 with changes

flhofer committed:
MKRWAN streamSkipUntil with timeout

flhofer committed:
MKRWAN added default values to waitResponse for error codes

flhofer committed:
MKRWAN waitReceive added binary parser for +RECVB=

flhofer committed:
MKRWAN -> hack for modem-Join put back with timed skip (for no-show +OK)

flhofer committed:
MKRWAN store and use msize

flhofer committed:
MKRWAN merged data receive blocks, add length check

flhofer committed:
MKRWAN fix power missing EQ, fix band-init inverted return

flhofer committed:
MKRWAN fix missed ADR constant

flhofer committed:
MKRWAN fix constants Mode vs Confirm status, added getter MsgConfirm

flhofer committed:
MKRWAN changed default compatibility on and test first, HC 64byte limit

flhofer committed:
MKRWAN replaced version() with compatibility-agnostic code

flhofer committed:
MKRWAN debug prints preamble ###

flhofer committed:
Standalone updater, add firmware 1.2.4 with TXPower set fix ab564f7

flhofer committed:
MKRWan add populateChannelsMask as a common refresher for get/setters

sslupsky reviewed (commented):
Added a few comments regarding outstanding issues I have come across related to this PR.

sslupsky commented on src/MKRWAN.h:
```diff
@@ -1047,79 +986,135 @@ class LoRaModem : public Stream
+			  (void)stream.readStringUntil(',').toInt();
+			  length = stream.readStringUntil('\r').toInt();
+			  (void)streamSkipUntil('\n');
+			  (void)streamSkipUntil('\n');
```
index should be 99 here?

flhofer replied:
The comment says so, but it seems the return code 99 has been removed in favor of buffer parsing + return state evaluation. For example, you have the modem in Class C. If you set a value and read from the modem, it could respond with `+RECV=<somemessage>+OK`. In this case, it pastes data to the buffer but still returns ok to the parameter set. You can poll if some new data arrived. If we would return 99, we get an error for the set as we didn't read the expected `+OK` yet, or we prefer the _receive_ to the ok.

sslupsky commented on src/MKRWAN.h:
```diff
@@ -1047,79 +986,135 @@ class LoRaModem : public Stream
+				  length = 0;
+				  continue;
+			  }
+			  if (data.endsWith(AT_RECVB)){ // Binary receive
```
Should return value (index) be a unique value for this message type?

flhofer replied:
Same as above..
Don't know how we might implement this in an other way.

sslupsky commented on src/MKRWAN.h:
```diff
@@ -1047,79 +986,135 @@ class LoRaModem : public Stream
+					}
+				  }
+			  }
+			  else	// String receive
```
return value should be unique?

flhofer replied:
Same as above..
Don't know how we might implement this in an other way.

flhofer replied:
I checked the history and at some point, the 99 return value has been removed. I think this was because an independent receive over RX2 initiated by the server could cause an unknown send status. Ex, duty cycle exceeded RX2 works, but send does not. However, with 99 as return status, you would never know.
If a data receive was successful can also be tested by `.available()`; same for ack, can be tested by `getMsgConfirmed()`. Both do not trigger another TX/RX.

I think we can close this.

flhofer committed:
Mkrwan waitresponse undandled data issue

flhofer commented on src/MKRWAN.h:
```diff
@@ -997,23 +940,19 @@ class LoRaModem : public Stream
-    }
-    streamSkipUntil('=');
-    return stream.readStringUntil('\r').toInt();
+  bool getJoinStatus() {
```
I think `connected` does the trick here. It can also be used as operator

flhofer replied:
Ok, the last commit changes it to public.

flhofer commented on src/MKRWAN.h:
```diff
@@ -452,23 +523,18 @@ class LoRaModem : public Stream
-        return false;
-    }
-    return true;
+    return setValue(GF(AT_CLASS), (char)_class);
```
Technically, class C is implemented, i.e., the modem stays in listen mode and you have to poll it to see if new data arrived. Didn't try it yet, though. Don't know if I get to use it in the following days. Still working on the test scripts
Class B is not implemented.

sslupsky replied:
Class C support requires that we consider that the samd21 might be sleeping between downlinks.  If that is the case, there may be data loss.  Couple ways to avoid it are to ensure SERCOM is properly configured to wake or do not send data until polled.

flhofer replied:
> Class C support requires that we consider that the samd21 might be sleeping between downlinks.
Class C is an always-on mode, thought for line-powered devices. I don't think that MCU sleep will be considered in such application cases. Am I missing out on something here 🤔

sebromero replied:
@flhofer You got a point. Imo when Class C is desired, we can live without sleep modes. However, if we can put the MCU to sleep while the modem keeps doing its job, we could save some energy which is always desirable.

sebromero replied:
@flhofer The polling mechanism which is currently implemented uses Class A. It sends a dummy message to open the RX1 and RX2 receive windows.

flhofer replied:
Every device has to support class A. Actually, the library right now always fills the rx-buffer if it gets something on the `waitresponse` method. However, wouldn't know how that behaves if you don't read the serial for seconds, if not minutes. The library does not support interrupt activation as for example does the TTN uno version. That should be the next thing to work on. This way you can put the MCU to sleep and wake-up on receive. The actual implementation is very basic and unfortunately relies completely on polls.
As an in-between thing you can just periodically read some value (eg. FCD) and then see if there is some data `.available`. Just a suggestion

flhofer committed:
Add support for +MSIZE for >=1.2.4, return value for poll()

flhofer committed:
Change getJoinStatus public

flhofer commented:
Hi! The last commit contains some minor fixes to use the updates on the firmware for 242 HEX bytes send and receive. It should also fix the DR0 stuck data rate for non-ADR send. The first tests seem ok. Once this all is through, I would close the pull request and end the work on this firmware version. (https://github.com/arduino/mkrwan1300-fw/pull/35)

@sslupsky @giulcioffi Any suggestions before we close 1.2.4 and switch to a new firmware?

flhofer committed:
Change getJoinStatus public - (cleanup compile error duplicate)

flhofer committed:
Update getChannelmask, resolve ambiguous NRVO return (could cause leak)

flhofer committed:
JoinABP add 1sec wait delay as for OTAA

flhofer committed:
SendMask add String reserve to avoid repeating memory reallocation

flhofer committed:
Power, fix return message parsing waiting for wrong answer

flhofer committed:
Added escaping (GF) for constants - compatibility

flhofer committed:
power fix erroneous return value reset for compatibilty

Highlighted unused port in waitResponse/receive

flhofer committed:
Updated data types for setters FCU FCD to u32

flhofer committed:
MKRWAN fix modemSend binary to use hex-translated bytes

flhofer committed:
Update example standalone to FW da2917c

flhofer committed:
Mkrwan add tracking of ADR status to keep last max-length value

flhofer committed:
Mkrwan simplified hex translation bin send

flhofer committed:
Mkrwan fix waitResponse wrong char read after timeout

flhofer commented:
Finished tests, think I'm done here. Anything else?

giulcioffi commented on src/MKRWAN.h:
```diff
@@ -671,22 +708,21 @@ class LoRaModem : public Stream
-    sendAT(GF("+REBOOT"));
-    if (waitResponse(10000L, "+EVENT=0,0") != 1) {
+    sendAT(GF(AT_RESET));
+    if (waitResponse(10000L, GF(AT_EVENT) GF(AT_EQ) "0,0") != 1) {
```
Hello @flhofer, I'm having a failure when calling `modem.restart()`: `waitResponse()` is timing out, probably because of a not handled case.
Could you please double check it on your side? Thanks!

flhofer replied:
Hi @giulcioffi 
sorry for the late reply. Found and fixed the reset error. My tests showed an unhandled situation in the library. Here is the log from my device with debug output.
```
Reset
### AT: 
### Data string: +OK
### AT: +REBOOT
### Data string: +EVENT          <-- valid terminator '=', but string does not match '+EVENT=0,0'
### Unhandled: +EVENT=0,0   <-- parsing timeout. we didn't receive any of the valid terminators (=,+,\r)
### AT: 
### Data string: +EVENT
### Data string: +EVENT=0,0
### Data string: +EVENT=0,0+OK.   <-- AT triggered new parsing.. now reacting to +OK
### AT: +DEV ?
### Data string: +DEV
### AT: +VER ?
```
Indeed, the library's waitResponse only parses `["=", "+", "\r"]` but the event message did not terminate with either of them. Thus, the check in the library timed out. I now added the `\r` to the reboot event message. Tested and works fine.
😃 ready to merge?

flhofer committed:
Merge branch 'master' into fix-hup1.2.3

flhofer committed:
Update Standalone firmware updater file 1.2.4 with MRKWAN FW fix e49d8e0

flhofer commented:
@sslupsky anyone reviewing this? would solve a lot of open issues 😃

nextstep79 commented:
> @sslupsky anyone reviewing this? would solve a lot of open issues 😃

Agree - looking forward to this being merged  👍

flhofer commented:
guys?

flhofer commented:
> guys?
> @giulcioffi @sslupsky @sebromero

Ping?

Larshand92 commented:
This PR fixes a lot of issues with the current master (v1.2.3), but it unfortunately seems to have introduced another. 

I'm not able to send binary data, or at least send it correctly (it works on master). For example, sending a package with 3 bytes [0,1,255] (0x00, 0x01, 0xFF) returns -7 (LORA_ERROR_RX). I have tried using `modem.write()` and `modem.print()`. `modem.write()` only works with single char/byte/int etc. and not for arrays. Trying to call `modem.write()` more than once give the same error -7. The only way i was able to send the binary data was with `modem.print()` and the `formanBin` flag set to `true`, but the formatting doubles the package size. Looking into the `modemSend()` code I have found the issue, every HEX (4 bits) is using 8 bits (1 byte) of TX buffer. The `endPackage()` returns the correct number of bytes, but the received package in the network server shows the double amount of bytes. I looks like this is a firmware bug, since I have not  been able to fix this in the `MKRWAN.h` code.

For my applications this is a big deal since I need to send as much data as possible as often as possible. 
I would love to use this branch, but until this is fixed I'm forced to use master with the issues it has.
